Source/engine/random.hpp

```diff
@@ -136,6 +136,181 @@ class DiabloGenerator {
+	}
+
+	void generate(uint32_t *begin, const uint32_t *end)
+	{
+		while (begin != end) {
+			*begin = next();
+			++begin;
+		}
+	}
+};
+
+// Adapted from https://prng.di.unimi.it/splitmix64.c written in 2015 by Sebastiano Vigna
+//
+// See also:
+//  Guy L. Steele, Doug Lea, and Christine H. Flood. 2014.
+//  Fast splittable pseudorandom number generators. SIGPLAN Not. 49, 10 (October 2014), 453–472.
+//  https://doi.org/10.1145/2714064.2660195
+class SplitMix64 {
+	uint64_t state;
+
+public:
+	SplitMix64(uint64_t state)
+	    : state(state)
+	{
+	}
+
+	uint64_t next()
+	{
+		uint64_t z = (state += 0x9e3779b97f4a7c15);
+		z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+		z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+		return z ^ (z >> 31);
+	}
+
+	void generate(uint64_t *begin, const uint64_t *end)
+	{
+		while (begin != end) {
+			*begin = next();
+			++begin;
+		}
+	}
+};
+
+/** Adapted from https://prng.di.unimi.it/xoshiro128plusplus.c written in 2019 by David Blackman and Sebastiano Vigna */
+class xoshiro128plusplus {
+public:
+	typedef uint32_t state[4];
+
+	xoshiro128plusplus() { seed(); }
+	xoshiro128plusplus(const state &s) { copy(this->s, s); }
+	xoshiro128plusplus(uint64_t initialSeed) { seed(initialSeed); }
+	xoshiro128plusplus(uint32_t initialSeed) { seed(initialSeed); }
+
+	uint32_t next()
+	{
+		const uint32_t result = rotl(s[0] + s[3], 7) + s[0];
+
+		const uint32_t t = s[1] << 9;
+
+		s[2] ^= s[0];
+		s[3] ^= s[1];
+		s[1] ^= s[2];
+		s[0] ^= s[3];
+
+		s[2] ^= t;
+
+		s[3] = rotl(s[3], 11);
+
+		return result;
+	}
+
+	/* This is the jump function for the generator. It is equivalent
+	   to 2^64 calls to next(); it can be used to generate 2^64
+	   non-overlapping subsequences for parallel computations. */
+	void jump()
+	{
+		static const uint32_t JUMP[] = { 0x8764000b, 0xf542d2d3, 0x6fa035c3, 0x77f2db5b };
+
+		uint32_t s0 = 0;
+		uint32_t s1 = 0;
+		uint32_t s2 = 0;
+		uint32_t s3 = 0;
+		for (int i = 0; i < sizeof JUMP / sizeof *JUMP; i++)
+			for (int b = 0; b < 32; b++) {
+				if (JUMP[i] & UINT32_C(1) << b) {
+					s0 ^= s[0];
+					s1 ^= s[1];
+					s2 ^= s[2];
+					s3 ^= s[3];
+				}
+				next();
+			}
+
+		s[0] = s0;
+		s[1] = s1;
+		s[2] = s2;
+		s[3] = s3;
+	}
+
+	void save(state &s) const
+	{
+		copy(s, this->s);
+	}
+
+private:
+	state s;
+
+	void seed(uint64_t value)
+	{
+		uint64_t seeds[2];
+		SplitMix64 seedSequence { value };
+		seedSequence.generate(seeds, seeds + 2);
+
+		s[0] = static_cast<uint32_t>(seeds[0] >> 32);
+		s[1] = static_cast<uint32_t>(seeds[0]);
+		s[2] = static_cast<uint32_t>(seeds[1] >> 32);
+		s[3] = static_cast<uint32_t>(seeds[1]);
+	}
+
+	void seed(uint32_t value)
+	{
+		SplitMix32 seedSequence { value };
+		seedSequence.generate(s, s + 4);
+	}
+
+	void seed()
+	{
+		uint64_t ts = timeSeed();
+		s[0] = static_cast<uint32_t>(ts >> 32);
+		s[1] = static_cast<uint32_t>(ts);
+
+		static std::random_device rd;
+		std::uniform_int_distribution<uint32_t> dist;
+		s[2] = dist(rd);
+		s[3] = dist(rd);
```

Comment:
Seems like convention is to mix the seed sources when an individual source is in doubt, how about using `timeSeed()` to initialise the state and then doing something like `for (uint32_t &sCell : s) s ^= dist(rd);`? Odds are that using time on its own is good enough™ but it shouldn't make much difference pulling an extra 64 bits from the random device. makes the code shorter as a bonus :D

Reply:
For the record, I hesitate to mix the seed sources using our own computations compared to what xoshiro can do on its own. IMO, many programmers will just use xor to "artfully" combine values without really considering the potential consequences. My main worry would be if a platform uses the system clock as a source of entropy for `std::random_device`, because xor of a value with itself will produce zero. This type of "resonance" between sources of entropy can easily introduce bias.

Normally, I might have suggested using a simple mx+b hashing method to mix the seed sources, but I went with your xor suggestion after all because I believe that SplitMix64 should be at least as effective as hashing to help ensure we don't introduce a bias.